src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,383 @@
+  (`nonsquare_iff_jacobi_sym_eq_neg_one`); in particular, in this case `a` is a
+  square mod `p` when the symbol has the value `1` (`is_square_of_jacobi_sym_eq_one`).
+
+* Quadratic reciprocity (`jacobi_sym_quadratic_reciprocity`,
+  `jacobi_sym_quadratic_reciprocity_one_mod_four`,
+  `jacobi_sym_quadratic_reciprocity_three_mod_four`)
+
+* The supplementary laws for `a = -1`, `a = 2`, `a = -2` (`jacobi_sym_neg_one`, `jacobi_sym_two`,
+  `jacobi_sym_neg_two`)
+
+* The symbol depends on `a` only via its residue class mod `b` (`jacobi_sym_mod_left`)
+  and on `b` only via its residue class mod `4*a` (`jacobi_sym_mod_right`)
+
+## Notations
+
+We define the notation `[a | b]ⱼ` for `jacobi_sym a b`, localized to `number_theory_symbols`.
+
+## Tags
+Jacobi symbol, quadratic reciprocity
+-/
+
+section jacobi
+
+/-!
+### Definition of the Jacobi symbol
+
+We define the Jacobi symbol $\Bigl\frac{a}{b}\Bigr$ for integers `a` and natural numbers `b` as the
+product of the Legendre symbols $\Bigl\frac{a}{p}\Bigr$, where `p` runs through the prime divisors
+(with multiplicity) of `b`, as provided by `b.factors`. This agrees with the Jacobi symbol
+when `b` is odd and gives less meaningful values when it is not (e.g., the symbol is `1`
+when `b = 0`). This is called `jacobi_sym a b`.
+
+We define localized notation (locale `number_theory_symbols`) `[a | b]ⱼ` for the Jacobi
+symbol `jacobi_sym a b`. (Unfortunately, there is no subscript "J" in unicode.)
+-/
+
+open zmod nat
+
+/-- The Jacobi symbol of `a` and `b` -/
+-- Since we need the fact that the factors are prime, we use `list.pmap`.
+def jacobi_sym (a : ℤ) (b : ℕ) : ℤ :=
+(b.factors.pmap (λ p pp, @legendre_sym p ⟨pp⟩ a) (λ p pf, prime_of_mem_factors pf)).prod
+
+-- Notation for the Jacobi symbol.
+localized "notation `[` a ` | ` b `]ⱼ` := jacobi_sym a b" in number_theory_symbols
+
+-- After this definition, list notation does not work in some contexts. Is this a bug?
+-- Specifying a precedence does not seem to help.
+-- (try #check [0] here)
+-- (It does work in tactic mode, at least.)
+
+/-!
+### Properties of the Jacobi symbol
+-/
+
+/-- The symbol `[a | 0]ⱼ` has the value `1`. -/
+@[simp] lemma jacobi_sym_zero_right (a : ℤ) : [a | 0]ⱼ = 1 :=
+by simp only [jacobi_sym, factors_zero, list.prod_nil, list.pmap]
+
+/-- The symbol `[a | 1]ⱼ` has the value `1`. -/
+@[simp] lemma jacobi_sym_one_right (a : ℤ) : [a | 1]ⱼ = 1 :=
+by simp only [jacobi_sym, factors_one, list.prod_nil, list.pmap]
+
+/-- The Legendre symbol `legendre_sym p a` with an integer `a` and a prime number `p`
+is the same as the Jacobi symbol `[a | p]ⱼ`. -/
+lemma legendre_sym.to_jacobi_sym {p : ℕ} [fp : fact p.prime] {a : ℤ} :
+  legendre_sym p a = [a | p]ⱼ :=
+by simp only [jacobi_sym, factors_prime fp.1, list.prod_cons, list.prod_nil, mul_one, list.pmap]
+
+/-- The Jacobi symbol is multiplicative in its second argument. -/
+lemma jacobi_sym_mul_right (a : ℤ) (b₁ b₂ : ℕ) [ne_zero b₁] [ne_zero b₂] :
```

Comment:
I think it is useful to have a similar lemma with the assumptions `b₁ ≠ 0` and `b₂ ≠ 0` explicit (without using the class `ne_zero`). Similarly below.

You can call the new lemma `jacobi_sym_mul_right'`, or maybe this one should be the primed version, I am not sure which one is more likely to be used.

Reply:
I think the present form is usually more convenient, since type class inference allows to deduce the non-vanishing assumption(s) automatically in some cases (e.g., when one argument is a positive numeral or a power of something known to be zero). So I'll call the new lemma `jacobi_sym_mul_right'`.